classes/class-wc-connect-payment-methods-store.php

```diff
@@ -0,0 +1,128 @@
+
+			$this->service_settings_store = $service_settings_store;
+			$this->api_client             = $api_client;
+			$this->logger                 = $logger;
+
+		}
+
+		public function fetch_payment_methods_from_connect_server() {
+
+			$response_body = $this->api_client->get_payment_methods();
+
+			if ( is_wp_error( $response_body ) ) {
+				$this->logger->log( $response_body, __FUNCTION__ );
+				return;
+			}
+
+			$payment_methods = $this->get_payment_methods_from_response_body( $response_body );
+			if ( is_wp_error( $payment_methods ) ) {
+				$this->logger->log( $payment_methods, __FUNCTION__ );
+				return;
+			}
+
+			$this->logger->log( 'Successfully loaded payment methods from server response.', __FUNCTION__ );
+			$this->update_last_fetch_timestamp();
+
+			// If we made it this far, it is safe to store the object
+			$this->update_payment_methods( $payment_methods );
+
+			// Potentially auto-update the selected method
+			$this->potentially_update_selected_payment_method_from_payment_methods( $payment_methods );
+		}
+
+		protected function potentially_update_selected_payment_method_from_payment_methods( $payment_methods ) {
+			$payment_method_count = count( $payment_methods );
+
+			// No payment methods at all? Clear anything we have stored
+			if ( 0 === $payment_method_count ) {
+				$this->service_settings_store->set_selected_payment_method_id( 0 );
+				return;
+			}
+
+			$payment_method_ids = array();
+			foreach ( (array) $payment_methods as $payment_method ) {
+				$payment_method_id = intval( $payment_method->payment_method_id );
+				if ( 0 !== $payment_method_id ) {
+					$payment_method_ids[] = $payment_method_id;
+				}
+			}
+
+			// Just one? Go ahead and use it
+			if ( 1 === $payment_method_count ) {
+				$this->service_settings_store->set_selected_payment_method_id( $payment_method_ids[ 0 ] );
+				return;
+			}
+
+			// Is the stored method id not in the list? Select the first one
+			$selected_payment_method_id = $this->service_settings_store->get_selected_payment_method_id();
+			if ( ! in_array( $selected_payment_method_id, $payment_method_ids ) ) {
+				$this->service_settings_store->set_selected_payment_method_id( $payment_method_ids[ 0 ] );
+			}
+		}
+
+		public function get_payment_methods() {
+			return get_option( 'wc_connect_payment_methods', array() );
+		}
+
+		protected function update_payment_methods( $payment_methods ) {
+			update_option( 'wc_connect_payment_methods', $payment_methods );
+		}
+
+		public function get_last_fetch_timestamp() {
+			return get_option( 'wc_connect_payment_methods_last_update', null );
+		}
+
+		protected function update_last_fetch_timestamp() {
+			update_option( 'wc_connect_payment_methods_last_update', time() );
+		}
+
+		protected function get_payment_methods_from_response_body( $response_body ) {
+			if ( ! is_object( $response_body ) ) {
+				return new WP_Error( 'payment_method_response_body_type', 'Expected but did not receive object for response body.' );
```

Comment:
Are these errors going to be presented to the user (in a notice)? Because they look like errors from a test case, not very user-friendly.